crowbar_framework/app/controllers/application_controller.rb

```diff
@@ -33,6 +33,9 @@ class ApplicationController < ActionController::Base
     Crowbar::Installer.successful? || \
     Rails.env.test?
   }
+  before_filter :upgrade, if: proc {
+    File.exist?("/var/lib/crowbar/upgrade/progress.yml")
```

Comment:
Also, to answer those questions:

1. if it were missing, it's considered that the upgrade is not running and any API can go through
2. Good question indeed.
3. When the `start_step` is called for the first timr
4. Yes